File: internal/confirm/confirm.go
// Package confirm implements the gap between a human approving an action and
// that action happening.
//
// The bug this package exists to prevent: approving a description and then
// executing something else. If the arguments are re-supplied by the agent
// after approval, the approval covered nothing. Here the arguments are frozen
// at the moment the intent is created, and execution uses the frozen copy.
package confirm

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/saadbutt/toolgate/internal/tools"
)

var (
	// ErrNotFound means no pending intent has that id.
	ErrNotFound = errors.New("confirm: no such intent")
	// ErrExpired means the approval window closed.
	ErrExpired = errors.New("confirm: intent expired")
	// ErrAlreadyUsed means the token was presented twice.
	ErrAlreadyUsed = errors.New("confirm: intent already used")
	// ErrBadToken means the token did not match.
	ErrBadToken = errors.New("confirm: token mismatch")
	// ErrWrongApprover means someone other than the intended approver tried
	// to confirm.
	ErrWrongApprover = errors.New("confirm: principal may not approve this intent")
	// ErrArgsMismatch means the arguments presented at execution differ from
	// the ones that were approved.
	ErrArgsMismatch = errors.New("confirm: arguments do not match what was approved")
)

// Intent is a consequential action waiting for a human.
type Intent struct {
	ID string
	// Tool and Args are the frozen request. Args is the copy that executes.
	Tool string
	Args tools.Args
	// ArgsHash is over the canonical encoding of Args, and is what makes
	// tampering detectable rather than merely unlikely.
	ArgsHash string
	// Summary is the sentence a human actually reads before approving.
	Summary string
	// RequestedBy is the agent that asked. Approver is who may say yes.
	RequestedBy string
	Approver    string
	CreatedAt   time.Time
	ExpiresAt   time.Time

	token []byte
	used  bool
}

// Store holds pending intents. Safe for concurrent use.
type Store struct {
	mu      sync.Mutex
	intents map[string]*Intent
	ttl     time.Duration
	now     func() time.Time
}

// NewStore returns a Store whose intents expire after ttl.
//
// A short ttl is part of the security model, not an ergonomic detail. An
// approval that stays valid for a day is an approval for a situation that no
// longer exists.
func NewStore(ttl time.Duration) *Store {
	if ttl <= 0 {
		ttl = 5 * time.Minute
	}
	return &Store{
		intents: make(map[string]*Intent),
		ttl:     ttl,
		now:     time.Now,
	}
}

// SetClock replaces the time source. Tests use it; production does not.
func (s *Store) SetClock(f func() time.Time) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.now = f
}

// Create freezes a request and returns the intent plus its one-time token.
//
// The token is returned once and never stored in recoverable form. Reading
// the store does not let you approve anything.
func (s *Store) Create(toolName string, args tools.Args, summary, requestedBy, approver string) (*Intent, string, error) {
	hash, err := HashArgs(args)
	if err != nil {
		return nil, "", err
	}

	idRaw := make([]byte, 12)
	if _, err := rand.Read(idRaw); err != nil {
		return nil, "", fmt.Errorf("confirm: generating id: %w", err)
	}
	tokRaw := make([]byte, 32)
	if _, err := rand.Read(tokRaw); err != nil {
		return nil, "", fmt.Errorf("confirm: generating token: %w", err)
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	now := s.now()

	in := &Intent{
		ID:          hex.EncodeToString(idRaw),
		Tool:        toolName,
		Args:        cloneArgs(args),
		ArgsHash:    hash,
		Summary:     summary,
		RequestedBy: requestedBy,
		Approver:    approver,
		CreatedAt:   now,
		ExpiresAt:   now.Add(s.ttl),
		token:       tokRaw,
	}
	s.intents[in.ID] = in
	return in, hex.EncodeToString(tokRaw), nil
}

// Pending returns an intent for display without consuming it.
func (s *Store) Pending(id string) (*Intent, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	in, ok := s.intents[id]
	if !ok {
		return nil, ErrNotFound
	}
	cp := *in
	cp.Args = cloneArgs(in.Args)
	cp.token = nil
	return &cp, nil
}

// Confirm consumes the intent and returns the frozen arguments.
//
// Everything that could go wrong is checked before the intent is marked used,
// and the intent is marked used before the arguments are returned, so a
// concurrent second attempt loses.
func (s *Store) Confirm(id, token, approver string) (string, tools.Args, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	in, ok := s.intents[id]
	if !ok {
		return "", nil, ErrNotFound
	}
	if in.used {
		return "", nil, ErrAlreadyUsed
	}
	if s.now().After(in.ExpiresAt) {
		return "", nil, ErrExpired
	}
	if approver != in.Approver {
		return "", nil, ErrWrongApprover
	}
	raw, err := hex.DecodeString(token)
	if err != nil || subtle.ConstantTimeCompare(raw, in.token) != 1 {
		return "", nil, ErrBadToken
	}

	in.used = true
	return in.Tool, cloneArgs(in.Args), nil
}

// VerifyArgs reports whether args are byte-identical to what was approved.
//
// The gate does not need this, because it executes the frozen copy. It exists
// so a caller that carries arguments alongside a token can prove they were not
// swapped in transit, and so the demo can show the tampering case failing.
func (s *Store) VerifyArgs(id string, args tools.Args) error {
	s.mu.Lock()
	in, ok := s.intents[id]
	s.mu.Unlock()
	if !ok {
		return ErrNotFound
	}
	h, err := HashArgs(args)
	if err != nil {
		return err
	}
	if subtle.ConstantTimeCompare([]byte(h), []byte(in.ArgsHash)) != 1 {
		return ErrArgsMismatch
	}
	return nil
}

// Reap deletes expired intents and reports how many went.
func (s *Store) Reap() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	now := s.now()
	n := 0
	for id, in := range s.intents {
		if now.After(in.ExpiresAt) {
			delete(s.intents, id)
			n++
		}
	}
	return n
}

// HashArgs returns a stable hash over the canonical encoding of args.
//
// encoding/json sorts map keys, so the encoding is deterministic for the
// value types the tools package permits. That determinism is load-bearing:
// if two encodings of the same arguments could differ, a legitimate
// confirmation would sometimes fail and the check would get disabled.
func HashArgs(args tools.Args) (string, error) {
	b, err := json.Marshal(map[string]any(args))
	if err != nil {
		return "", fmt.Errorf("confirm: encoding args: %w", err)
	}
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:]), nil
}

func cloneArgs(in tools.Args) tools.Args {
	out := make(tools.Args, len(in))
	for k, v := range in {
		out[k] = v
	}
	return out
}

File: typescript/src/confirm.ts
/**
 * The gap between a human approving an action and that action happening.
 *
 * The bug this exists to prevent: approving a description and executing
 * something else. If the arguments are re-supplied after approval, the
 * approval covered nothing. Here they are frozen when the intent is created,
 * and `confirm` hands back the frozen copy.
 *
 * This is a port of internal/confirm from the Go implementation, kept to the
 * one piece most likely to be got wrong elsewhere. The Go version is the
 * complete system; this is the part worth having in your own stack.
 */

import { randomBytes, timingSafeEqual } from "node:crypto";
import { type Args, hashArgs } from "./canonical.ts";

export type ConfirmErrorCode =
  | "not_found"
  | "expired"
  | "already_used"
  | "bad_token"
  | "wrong_approver"
  | "args_mismatch";

export class ConfirmError extends Error {
  // Assigned in the body rather than as a constructor parameter property,
  // so the file runs under `node --experimental-strip-types` with no build
  // step. Keeping that true is worth more than the shorter syntax.
  readonly code: ConfirmErrorCode;

  constructor(code: ConfirmErrorCode, message: string) {
    super(message);
    this.name = "ConfirmError";
    this.code = code;
  }
}

/** A consequential action waiting for a human. */
export interface Intent {
  readonly id: string;
  readonly tool: string;
  /** The frozen, validated arguments. This copy is what executes. */
  readonly args: Args;
  /** Hash over the canonical encoding, which makes tampering detectable. */
  readonly argsHash: string;
  /** The sentence a human actually reads before approving. */
  readonly summary: string;
  readonly requestedBy: string;
  readonly approver: string;
  readonly createdAt: Date;
  readonly expiresAt: Date;
}

interface StoredIntent extends Intent {
  token: Buffer;
  used: boolean;
}

export interface StoreOptions {
  /**
   * How long an approval stays valid.
   *
   * Short by design. An approval that survives a day is an approval for a
   * situation that no longer exists.
   */
  ttlMs?: number;
  /** Injectable clock, for tests. */
  now?: (() => Date) | undefined;
}

export class IntentStore {
  readonly #intents = new Map<string, StoredIntent>();
  readonly #ttlMs: number;
  readonly #now: () => Date;

  constructor(opts: StoreOptions = {}) {
    this.#ttlMs = opts.ttlMs ?? 5 * 60_000;
    this.#now = opts.now ?? (() => new Date());
  }

  /**
   * Freezes a request and returns the intent plus a one-time token.
   *
   * The token is returned once and never stored in recoverable form, so
   * reading the store does not let you approve anything.
   */
  create(
    tool: string,
    args: Args,
    summary: string,
    requestedBy: string,
    approver: string,
  ): { intent: Intent; token: string } {
    const now = this.#now();
    const stored: StoredIntent = {
      id: randomBytes(12).toString("hex"),
      tool,
      // Copied, so a caller that mutates its own object afterwards cannot
      // change what was approved.
      args: { ...args },
      argsHash: hashArgs(args),
      summary,
      requestedBy,
      approver,
      createdAt: now,
      expiresAt: new Date(now.getTime() + this.#ttlMs),
      token: randomBytes(32),
      used: false,
    };
    this.#intents.set(stored.id, stored);
    return { intent: this.#public(stored), token: stored.token.toString("hex") };
  }

  /** Returns an intent for display without consuming it. */
  pending(id: string): Intent {
    const found = this.#intents.get(id);
    if (!found) throw new ConfirmError("not_found", `no intent ${id}`);
    return this.#public(found);
  }

  /**
   * Consumes the intent and returns the frozen arguments.
   *
   * Every failure is checked before the intent is marked used, and it is
   * marked used before the arguments are returned. Node is single-threaded
   * per event-loop turn and there is no await inside this method, so a second
   * caller cannot interleave and win the same token.
   */
  confirm(id: string, token: string, approver: string): { tool: string; args: Args } {
    const found = this.#intents.get(id);
    if (!found) throw new ConfirmError("not_found", `no intent ${id}`);
    if (found.used) throw new ConfirmError("already_used", "intent already used");
    if (this.#now() > found.expiresAt) throw new ConfirmError("expired", "intent expired");
    if (approver !== found.approver) {
      throw new ConfirmError("wrong_approver", `${approver} may not approve this intent`);
    }

    let supplied: Buffer;
    try {
      supplied = Buffer.from(token, "hex");
    } catch {
      throw new ConfirmError("bad_token", "token is not valid hex");
    }
    // timingSafeEqual throws on a length mismatch, which would itself leak
    // length information through the exception path, so length is compared
    // first and both paths end in the same error.
    if (supplied.length !== found.token.length || !timingSafeEqual(supplied, found.token)) {
      throw new ConfirmError("bad_token", "token mismatch");
    }

    found.used = true;
    return { tool: found.tool, args: { ...found.args } };
  }

  /**
   * Reports whether args are byte-identical to what was approved.
   *
   * Execution does not need this, because it uses the frozen copy. It exists
   * for callers that carry arguments alongside a token and want to prove they
   * were not swapped in transit.
   */
  verifyArgs(id: string, args: Args): void {
    const found = this.#intents.get(id);
    if (!found) throw new ConfirmError("not_found", `no intent ${id}`);
    if (hashArgs(args) !== found.argsHash) {
      throw new ConfirmError("args_mismatch", "arguments do not match what was approved");
    }
  }

  /** Deletes expired intents and returns how many went. */
  reap(): number {
    const now = this.#now();
    let n = 0;
    for (const [id, intent] of this.#intents) {
      if (now > intent.expiresAt) {
        this.#intents.delete(id);
        n++;
      }
    }
    return n;
  }

  #public(s: StoredIntent): Intent {
    const { token, used, ...rest } = s;
    void token;
    void used;
    return { ...rest, args: { ...s.args } };
  }
}

/**
 * Approves an intent and runs the handler against the frozen arguments.
 *
 * The arguments are not a parameter. That is the whole design: the caller
 * cannot supply them, so the caller cannot change them between approval and
 * effect.
 */
export async function confirmAndRun<T>(
  store: IntentStore,
  id: string,
  token: string,
  approver: string,
  handlers: Record<string, (args: Args) => Promise<T> | T>,
): Promise<T> {
  const { tool, args } = store.confirm(id, token, approver);
  const handler = handlers[tool];
  if (!handler) throw new Error(`confirmAndRun: no handler registered for ${tool}`);
  return await handler(args);
}
